python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,189 @@
+    is a write operation being performed.
+
+    See more here :ref:`copy-on-write-dev-doc`.
+
+    Use the factory function `as_buffer` to create a CopyOnWriteBuffer
+    instance.
+    """
+
+    _instances: DefaultDict[Tuple, WeakSet] = defaultdict(WeakSet)
+    """This dict keeps track of all instances that have the same `ptr`
+    and `size` attributes.  Each key of the dict is a `(ptr, size)`
+    tuple and the corresponding value is a set of weak references to
+    instances with that `ptr` and `size`."""
+
+    # TODO: This is synonymous to SpillableBuffer._exposed attribute
+    # and has to be merged.
+    _zero_copied: bool
+
+    def _finalize_init(self):
+        self.__class__._instances[self._ptr].add(self)
+        self._instances = self.__class__._instances[self._ptr]
+        self._zero_copied = False
+        weakref.finalize(self, _keys_cleanup, self._ptr)
+
+    @classmethod
+    def _from_device_memory(
+        cls: Type[T], data: Any, *, exposed: bool = False
+    ) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+        exposed : bool, optional
+            Mark the buffer as zero copied.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._finalize_init()
+        ret._zero_copied = exposed
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @property
+    def _is_shared(self):
+        """
+        Return `True` if `self`'s memory is shared with other columns.
+        """
+        return len(self._instances) > 1
+
+    def get_ptr(self, mode: str = "write") -> int:
+        """Device pointer to the start of the buffer.
+
+        Parameters
+        ----------
+        mode : str, default 'write'
+            Supported values are {"read", "write"}
+            If "write", when weak-references exist, they
+            are unlinked and the data pointed to may be modified
+            by the caller. If "read", the data pointed to
+            must not be modified by the caller.
+            Failure to fulfill this contract will cause
+            incorrect behavior.
+
+        See Also
+        --------
+        Buffer.get_ptr
+        SpillableBuffer.get_ptr
+        """
+        if mode == "read":
+            return self._ptr
+        elif mode == "write":
+            self._unlink_shared_buffers()
+            return self._ptr
+        else:
+            raise ValueError(f"Incorrect mode passed : {mode}")
```

Comment:
FWIW you could simplify this with 
```suggestion
        if mode == "write":
            self._unlink_shared_buffers()
        elif mode != "read":
            raise ValueError(f"Incorrect mode passed : {mode}")
        return self._ptr
```
Just noting because it came to mind, not really necessary.